cpp/include/cuvs/neighbors/common.hpp

```diff
@@ -615,5 +616,56 @@ enable_if_valid_list_t<ListT> deserialize_list(const raft::resources& handle,
                                                const typename ListT::spec_type& store_spec,
                                                const typename ListT::spec_type& device_spec);
 }  // namespace ivf
+}  // namespace cuvs::neighbors
+
+namespace cuvs::neighbors {
+using namespace raft;
+
+template <typename AnnIndexType, typename T, typename IdxT>
+struct iface {
+  iface() : mutex_(std::make_shared<std::mutex>()) {}
+
+  const IdxT size() const { return index_.value().size(); }
+
+  std::optional<AnnIndexType> index_;
+  std::shared_ptr<std::mutex> mutex_;
+};
+
+template <typename AnnIndexType, typename T, typename IdxT, typename Accessor>
+void build(const raft::device_resources& handle,
+           cuvs::neighbors::iface<AnnIndexType, T, IdxT>& interface,
+           const cuvs::neighbors::index_params* index_params,
+           raft::mdspan<const T, matrix_extent<int64_t>, row_major, Accessor> index_dataset);
+
+template <typename AnnIndexType, typename T, typename IdxT, typename Accessor1, typename Accessor2>
+void extend(
+  const raft::device_resources& handle,
+  cuvs::neighbors::iface<AnnIndexType, T, IdxT>& interface,
+  raft::mdspan<const T, matrix_extent<int64_t>, row_major, Accessor1> new_vectors,
+  std::optional<raft::mdspan<const IdxT, vector_extent<int64_t>, layout_c_contiguous, Accessor2>>
+    new_indices);
+
+template <typename AnnIndexType, typename T, typename IdxT>
+void search(const raft::device_resources& handle,
```

Comment:
Do we need this function as part of the public API? Isn't this just used internally in iface.hpp?

Here we have a search API that takes `queries` as host_matrix, and returns `neighbors` as device matrix. I see that this interface is used internally in the implementation of the MG algorithm. In contrast the public API in mg.hpp defines methods that takes host `queries` and return host `neighbors`. I might be missing something, but I do not see the role of this in the public API.

Reply:
@cjnolet please confirm that this is the case, and that we indeed want a search interface with _host_ input queries and _device_ output.

Reply:
As discussed offline with @tfeher, the search function used in the mg namespace will not be exposed in the public `iface` declaration. Instead a search function with inputs and outputs on device will take place.